### PR TITLE
layers: Validate VK_EXT_pipeline_creation_feedback

### LINKED ISSUE #830: Add validation for VK_EXT_pipeline_creation_feedback
There's no coverage for the new VUs in VK_EXT_pipeline_creation_feedback - only implicit VUs are supported. Adding support for the handful of explicit VUs should be fairly straightforward, but it would be good to get coverage.

CC @jfroy as the spec author in case this is something that he's able to help with (or find someone who can!)

### PR DESCRIPTION
**Pull request by hysw**
Check valid usage for VUID-VkPipelineCreationFeedbackCreateInfoEXT-*.

This pull request fixes #830

### CODE REVIEW

**Comment by jfroy** [comment]
@jzulauf-lunarg Can you please review?

File: layers/parameter_validation_utils.cpp [hunk 1]
```diff
@@ -935,6 +935,17 @@ bool StatelessValidation::manual_PreCallValidateCreateGraphicsPipelines(VkDevice
                 }
             }
 
+            auto feedback_struct = lvl_find_in_chain<VkPipelineCreationFeedbackCreateInfoEXT>(pCreateInfos[i].pNext);
+            if (feedback_struct != nullptr && feedback_struct->pipelineStageCreationFeedbackCount != pCreateInfos[i].stageCount) {
+                skip |=
+                    log_msg(report_data, VK_DEBUG_REPORT_ERROR_BIT_EXT, VK_DEBUG_REPORT_OBJECT_TYPE_PIPELINE_EXT, VK_NULL_HANDLE,
+                            "VUID-VkPipelineCreationFeedbackCreateInfoEXT-pipelineStageCreationFeedbackCount-02668",
+                            "vkCreateGraphicsPipelines(): in pCreateInfo[%u], "
```
Comment (jzulauf-lunarg) [inline layers/parameter_validation_utils.cpp]: `i` is also `uint32_t` so `PRIu32` here as well

File: layers/parameter_validation_utils.cpp [hunk 2]
```diff
@@ -1764,6 +1775,14 @@ bool StatelessValidation::manual_PreCallValidateCreateComputePipelines(VkDevice
         skip |= validate_string("vkCreateComputePipelines",
                                 ParameterName("pCreateInfos[%i].stage.pName", ParameterName::IndexVector{i}),
                                 "VUID-VkPipelineShaderStageCreateInfo-pName-parameter", pCreateInfos[i].stage.pName);
+        auto feedback_struct = lvl_find_in_chain<VkPipelineCreationFeedbackCreateInfoEXT>(pCreateInfos[i].pNext);
+        if (feedback_struct != nullptr && feedback_struct->pipelineStageCreationFeedbackCount != 1) {
+            skip |= log_msg(report_data, VK_DEBUG_REPORT_ERROR_BIT_EXT, VK_DEBUG_REPORT_OBJECT_TYPE_PIPELINE_EXT, VK_NULL_HANDLE,
+                            "VUID-VkPipelineCreationFeedbackCreateInfoEXT-pipelineStageCreationFeedbackCount-02669",
+                            "vkCreateComputePipelines(): in pCreateInfo[%u], VkPipelineCreationFeedbackEXT"
```
Comment (jzulauf-lunarg) [inline layers/parameter_validation_utils.cpp]: `i` is also `uint32_t` so `PRIu32` here as well

File: layers/parameter_validation_utils.cpp [hunk 3]
```diff
@@ -1764,6 +1775,14 @@ bool StatelessValidation::manual_PreCallValidateCreateComputePipelines(VkDevice
         skip |= validate_string("vkCreateComputePipelines",
                                 ParameterName("pCreateInfos[%i].stage.pName", ParameterName::IndexVector{i}),
                                 "VUID-VkPipelineShaderStageCreateInfo-pName-parameter", pCreateInfos[i].stage.pName);
+        auto feedback_struct = lvl_find_in_chain<VkPipelineCreationFeedbackCreateInfoEXT>(pCreateInfos[i].pNext);
+        if (feedback_struct != nullptr && feedback_struct->pipelineStageCreationFeedbackCount != 1) {
```
Comment (jzulauf-lunarg) [inline layers/parameter_validation_utils.cpp]: style preference for parens around logically combined comparisons.

File: layers/parameter_validation_utils.cpp [hunk 4]
```diff
@@ -935,6 +935,17 @@ bool StatelessValidation::manual_PreCallValidateCreateGraphicsPipelines(VkDevice
                 }
             }
 
+            auto feedback_struct = lvl_find_in_chain<VkPipelineCreationFeedbackCreateInfoEXT>(pCreateInfos[i].pNext);
+            if (feedback_struct != nullptr && feedback_struct->pipelineStageCreationFeedbackCount != pCreateInfos[i].stageCount) {
```
Comment (jzulauf-lunarg) [inline layers/parameter_validation_utils.cpp]: style preference for parens around logically combined comparisons.

File: layers/parameter_validation_utils.cpp [hunk 5]
```diff
@@ -2991,6 +3010,28 @@ bool StatelessValidation::manual_PreCallValidateCreateAccelerationStructureNV(
...
+bool StatelessValidation::manual_PreCallValidateCreateRayTracingPipelinesNV(VkDevice device, VkPipelineCache pipelineCache,
+                                                                            uint32_t createInfoCount,
+                                                                            const VkRayTracingPipelineCreateInfoNV *pCreateInfos,
+                                                                            const VkAllocationCallbacks *pAllocator,
+                                                                            VkPipeline *pPipelines) {
+    bool skip = false;
+
+    for (uint32_t i = 0; i < createInfoCount; i++) {
+        auto feedback_struct = lvl_find_in_chain<VkPipelineCreationFeedbackCreateInfoEXT>(pCreateInfos[i].pNext);
+        if (feedback_struct != nullptr && feedback_struct->pipelineStageCreationFeedbackCount != pCreateInfos[i].stageCount) {
```
Comment (jzulauf-lunarg) [inline layers/parameter_validation_utils.cpp]: style preference for parens around logically combined comparisons.

File: layers/parameter_validation_utils.cpp [hunk 6]
```diff
@@ -2991,6 +3010,28 @@ bool StatelessValidation::manual_PreCallValidateCreateAccelerationStructureNV(
...
+            skip |= log_msg(report_data, VK_DEBUG_REPORT_ERROR_BIT_EXT, VK_DEBUG_REPORT_OBJECT_TYPE_PIPELINE_EXT, VK_NULL_HANDLE,
+                            "VUID-VkPipelineCreationFeedbackCreateInfoEXT-pipelineStageCreationFeedbackCount-02670",
+                            "vkCreateRayTracingPipelinesNV(): in pCreateInfo[%u], "
```
Comment (jzulauf-lunarg) [inline layers/parameter_validation_utils.cpp]: `i` is also `uint32_t` so `PRIu32` here as well

File: tests/layer_validation_tests.cpp [hunk 7]
```diff
@@ -39186,6 +39186,280 @@ TEST_F(VkLayerTest, CooperativeMatrixNV) {
...
+    VkPipelineCreationFeedbackEXT feedbacks[3] = {};
+    feedback_info.sType = VK_STRUCTURE_TYPE_PIPELINE_CREATION_FEEDBACK_CREATE_INFO_EXT;
+
+    feedback_info.pPipelineCreationFeedback = &feedbacks[0];
+    feedback_info.pipelineStageCreationFeedbackCount = 2;
+    feedback_info.pPipelineStageCreationFeedbacks = &feedbacks[1];
+
+    auto set_feedback = [&feedback_info](CreatePipelineHelper &helper) { helper.gp_ci_.pNext = &feedback_info; };
+
+    CreatePipelineHelper::OneshotTest(*this, set_feedback, VK_DEBUG_REPORT_ERROR_BIT_EXT,
```
Comment (jzulauf-lunarg) [inline tests/layer_validation_tests.cpp]: 👍 for using the helpers!

File: tests/layer_validation_tests.cpp [hunk 8]
```diff
@@ -39186,6 +39186,280 @@ TEST_F(VkLayerTest, CooperativeMatrixNV) {
...
+    }
+    ASSERT_NO_FATAL_FAILURE(InitState());
+    ASSERT_NO_FATAL_FAILURE(InitRenderTarget());
+
+    VkDescriptorSetObj descriptorSet(m_device);
+    descriptorSet.AppendDummy();
+    descriptorSet.CreateVKDescriptorSet(m_commandBuffer);
+
+    // Create a minimal compute pipeline
+    std::string cs_text = "#version 450\nvoid main() {}\n";  // minimal no-op shader
```
Comment (jzulauf-lunarg) [inline tests/layer_validation_tests.cpp]: const char * when string functionality isn't needed?

File: tests/layer_validation_tests.cpp [hunk 9]
```diff
@@ -39186,6 +39186,280 @@ TEST_F(VkLayerTest, CooperativeMatrixNV) {
...
+    if (DeviceExtensionSupported(gpu(), nullptr, VK_EXT_PIPELINE_CREATION_FEEDBACK_EXTENSION_NAME)) {
+        m_device_extension_names.push_back(VK_EXT_PIPELINE_CREATION_FEEDBACK_EXTENSION_NAME);
+    } else {
+        printf("%s Extension %s is not supported.\n", kSkipPrefix, VK_EXT_PIPELINE_CREATION_FEEDBACK_EXTENSION_NAME);
+        return;
+    }
+    ASSERT_NO_FATAL_FAILURE(InitState());
+    ASSERT_NO_FATAL_FAILURE(InitRenderTarget());
+
+    VkPipelineCreationFeedbackCreateInfoEXT feedback_info = {};
```
Comment (jzulauf-lunarg) [inline tests/layer_validation_tests.cpp]: `auto ... lvl_init_struct<...>` ?

File: tests/layer_validation_tests.cpp [hunk 10]
```diff
@@ -39186,6 +39186,280 @@ TEST_F(VkLayerTest, CooperativeMatrixNV) {
...
+        "void main() {                                  \n"
+        "    hitValue = 0.0;                            \n"
+        "}                                              \n";
+
+    VkShaderObj rgs(m_device, rayGenShaderText, VK_SHADER_STAGE_RAYGEN_BIT_NV, this);
+    VkShaderObj chs(m_device, closestHitShaderText, VK_SHADER_STAGE_CLOSEST_HIT_BIT_NV, this);
+    VkShaderObj mis(m_device, missShaderText, VK_SHADER_STAGE_MISS_BIT_NV, this);
+
+    VkPipelineShaderStageCreateInfo rayStages[3];
+    memset(&rayStages[0], 0, sizeof(rayStages));
```
Comment (jzulauf-lunarg) [inline tests/layer_validation_tests.cpp]: This looks like copy/paste from other ray tracing tests.  Can this refactor into a `Helper` `SafeSane` initializer(s)?

**Comment by jzulauf-lunarg** [comment]
this is a good contribution.  thanks.  style/cleanup requests only.

File: tests/layer_validation_tests.cpp [hunk 11]
```diff
@@ -1333,6 +1333,249 @@ struct CreatePipelineHelper {
...
+    // flags, error can be any args accepted by "SetDesiredFailure".
+    template <typename Test, typename OverrideFunc, typename Error>
+    static void OneshotTest(Test &test, OverrideFunc &info_override, const VkFlags flags, const std::vector<Error> &errors,
+                            bool positive_test = false) {
+        CreateNVRayTracingPipelineHelper helper(test);
+        helper.InitInfo();
+        info_override(helper);
+        helper.InitState();
+
+        for (const auto &error : errors) test.Monitor()->SetDesiredFailureMsg(flags, error);
```
Comment (jzulauf-lunarg) [inline tests/layer_validation_tests.cpp]: The code you based this one is probably incorrect.  The loop should be only executed when !positive.

```
if(postive)
    ... ExpectSuccess ...
else
    for loop.
```

(the code is some early stuff I wrote so, mea culpa here.)

File: tests/layer_validation_tests.cpp [hunk 12]
```diff
@@ -1333,6 +1333,249 @@ struct CreatePipelineHelper {
         OneshotTest(test, info_override, flags, std::vector<Error>(1, error), positive_test);
     }
 };
+
```
Comment (jzulauf-lunarg) [inline tests/layer_validation_tests.cpp]: This is a good looking helper.  @jeffbolznv  -- would this inspire/enable additional test coverage?

File: tests/layer_validation_tests.cpp [hunk 13]
```diff
@@ -39186,6 +39298,139 @@ TEST_F(VkLayerTest, CooperativeMatrixNV) {
...
+    pipeline_info.stage.pName = "main";
+    pipeline_info.stage.module = cs_obj.handle();
+
+    {
+        m_errorMonitor->SetDesiredFailureMsg(
+            VK_DEBUG_REPORT_ERROR_BIT_EXT, "VUID-VkPipelineCreationFeedbackCreateInfoEXT-pipelineStageCreationFeedbackCount-02669");
+        VkPipeline cs_pipeline = VK_NULL_HANDLE;
+        vkCreateComputePipelines(device(), VK_NULL_HANDLE, 1, &pipeline_info, nullptr, &cs_pipeline);
+        vkDestroyPipeline(device(), cs_pipeline, nullptr);
+        m_errorMonitor->VerifyNotFound();
```
[the remark this answers is not in the document]
Comment (jzulauf-lunarg) [reply tests/layer_validation_tests.cpp]: (and I note likely this derives from a bad exemplar)

File: tests/layer_validation_tests.cpp (hunk 11 above)
Comment (hysw) [reply tests/layer_validation_tests.cpp]: I modified function names/signatures in CreateNVRayTracingPipelineHelper, so it's less likely to cause further confusion.

**Comment by jzulauf-lunarg** [comment]
@hysw  -- we typically like layer and lest components in separate commits.  First a "layers: ..." commit, and then a "tests: ..." commit.

**Pull request merged**